spark/v3.2/spark/src/main/java/org/apache/iceberg/spark/Spark3Util.java

```diff
@@ -781,10 +801,29 @@ public static TableIdentifier identifierToTableIdentifier(Identifier identifier)
             Object value = CatalystTypeConverters.convertToScala(catalystValue, field.dataType());
             values.put(field.name(), String.valueOf(value));
           });
-          return new SparkPartition(values, partition.path().toString(), format);
+          FileStatus fileStatus =
+              scala.collection.JavaConverters.seqAsJavaListConverter(partition.files()).asJava().get(0);
+
+          return new SparkPartition(values, fileStatus.getPath().getParent().toString(), format);
         }).collect(Collectors.toList());
   }
 
+  private static List<org.apache.spark.sql.catalyst.expressions.Expression> getPartitionFilterExpressions(
+      SparkSession spark, String tableName, Map<String, String> partitionFilter) {
+    List<org.apache.spark.sql.catalyst.expressions.Expression> filterExpressions = Lists.newArrayList();
+    for (Map.Entry<String, String> entry : partitionFilter.entrySet()) {
+      String filter = entry.getKey() + " = '" + entry.getValue() + "'";
+      try {
+        org.apache.spark.sql.catalyst.expressions.Expression expression =
+            SparkExpressionConverter.collectResolvedSparkExpression(spark, tableName, filter);
```

Comment:
`collectResolvedSparkExpression` is really expensive. Why does this need to call it?

Can't this produce `Expression` instances directly rather than building strings and converting with fake plans?

Reply:
@rdblue Thank you very much for reviewing my PR on the weekend!

Do you mean constructing a filter `Expression` instead of letting Spark generate the `Expression`? I initially generated  `Expression` like this
```
        BoundReference ref = new BoundReference(index, dataType, true);
        switch (dataType.typeName()) {
          case "integer":
            filterExpressions.add(new EqualTo(ref,
                org.apache.spark.sql.catalyst.expressions.Literal.create(Integer.parseInt(entry.getValue()),
                DataTypes.IntegerType)));
            break;
```
There are some concerns from the reviewers because we need to test each of the data types. Then I changed the code to call `collectResolvedSparkExpression`.

I will address all the other comments after I find out what to do for this one, so I can fix all the problems in one commit.

Reply:
Yes, I think you should construct the filter expression directly rather than calling `collectResolvedSparkExpression`.

Reply:
@rdblue I changed the code to construct filter expression directly. Could you please take one more look? Thank you very much!

I have manually checked to make sure the filter expression can be created correctly for all the numeric types, Date and Timestamp. There are tests for String and int partition filters in `TestAddFilesProcedure`, and I added a test for Date partition filter. There is a bug with Timestamp partition filter in the current code. I will fix it in a separate PR.

The build failure doesn't seem to be related to my changes.